File: Hyeyoung/11주차/HeapSort.java
import java.util.*;
import java.io.*;

public class HeapSort {
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int[] arr = new int[n + 1];
        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++)
        {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        br.close();

        heapSort(arr, n);
    
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= n; i++)
        {
            sb.append(arr[i] + " ");
        }
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        bw.write(sb.toString());
        bw.close();
    }

    private static void heapSort(int[] arr, int n)
    {
        for (int i = n/2; i >= 1; i--)
        {
            heapify(arr, n, i);
        }

        for (int i = n; i > 1; i--)
        {
            swap(arr, i, 1);
            heapify(arr, i-1, 1);
        }
    }

    private static void heapify(int[] arr, int count, int idx)
    {
        int largest = idx;
        int leftChild = 2 * idx;
        int rightChild = 2 * idx + 1;

        if (leftChild <= count && arr[leftChild] > arr[largest])
        {
            largest = leftChild;
        }
        if (rightChild <= count && arr[rightChild] > arr[largest])
        {
            largest = rightChild;
        }

        if (largest != idx)
        {
            swap(arr, idx, largest);
            heapify(arr, count, largest);
        }
    }

    private static void swap(int[] arr, int idx1, int idx2)
    {
        int tmp = arr[idx1];
        arr[idx1] = arr[idx2];
        arr[idx2] = tmp;
    }
}

File: Yejin/Yejin/CodeTree/11th/HeapSort.swift
//
//  HeapSort.swift
//  Yejin
//
//  Created by Yejin Hong on 1/22/25.
//
// 첫번째 줄에는 n
// 두번째 줄에는 n개의 원소 값이 공백을 두고 주어짐
// 1 <= n <= 100,000
// 1 <= 원소 값 <= 100,000

func heapify(array: [Int], n: Int, i: Int) {
    var arr = array
    
    var largest = i
    var l = i * 2
    var r = i * 2 - 1
    
    if l <= n && array[l] > array[largest] {
        largest = l
    }
    
    if r <= n && array[r] > array[largest] {
        largest = r
    }
    
    if largest != i {
        // swap(arr[i], arr[largest])
        heapify(array: array, n: n, i: largest)
    }
}

func heapSort() {
    let number = readLine()
    let elements = readLine()!.split(separator: " ").map{ Int($0)! }
    
    
}
